price-feeder/oracle/provider/mexc.go

```diff
@@ -0,0 +1,504 @@
+		candles:         map[string][]types.CandlePrice{},
+		subscribedPairs: map[string]types.CurrencyPair{},
+	}
+
+	if err := provider.SubscribeCurrencyPairs(pairs...); err != nil {
+		return nil, err
+	}
+
+	go provider.handleWebSocketMsgs(ctx)
+
+	return provider, nil
+}
+
+// GetTickerPrices returns the tickerPrices based on the provided pairs.
+func (p *MexcProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]types.TickerPrice, error) {
+	tickerPrices := make(map[string]types.TickerPrice, len(pairs))
+
+	for _, cp := range pairs {
+		key := currencyPairToMexcPair(cp)
+		price, err := p.getTickerPrice(key)
+		if err != nil {
+			return nil, err
+		}
+		tickerPrices[cp.String()] = price
+	}
+
+	return tickerPrices, nil
+}
+
+// GetCandlePrices returns the candlePrices based on the provided pairs.
+func (p *MexcProvider) GetCandlePrices(pairs ...types.CurrencyPair) (map[string][]types.CandlePrice, error) {
+	candlePrices := make(map[string][]types.CandlePrice, len(pairs))
+
+	for _, cp := range pairs {
+		key := currencyPairToMexcPair(cp)
+		prices, err := p.getCandlePrices(key)
+		if err != nil {
+			return nil, err
+		}
+		candlePrices[cp.String()] = prices
+	}
+
+	return candlePrices, nil
+}
+
+// SubscribeCurrencyPairs subscribe all currency pairs into ticker and candle channels.
+func (p *MexcProvider) SubscribeCurrencyPairs(cps ...types.CurrencyPair) error {
+	if len(cps) == 0 {
+		return fmt.Errorf("currency pairs is empty")
+	}
+
+	if err := p.subscribeChannels(cps...); err != nil {
+		return err
+	}
+
+	p.setSubscribedPairs(cps...)
+	return nil
+}
+
+// subscribeChannels subscribe to the ticker and candle channels for all currency pairs.
+func (p *MexcProvider) subscribeChannels(cps ...types.CurrencyPair) error {
+	if err := p.subscribeTickers(cps...); err != nil {
+		return err
+	}
+
+	return p.subscribeCandles(cps...)
+}
+
+// subscribeTickers subscribe to the ticker channel for all currency pairs.
+func (p *MexcProvider) subscribeTickers(cps ...types.CurrencyPair) error {
+	pairs := make([]string, len(cps))
+
+	for i, cp := range cps {
+		pairs[i] = currencyPairToMexcPair(cp)
+	}
+
+	return p.subscribePairs(pairs...)
+}
+
+// subscribeCandles subscribe to the candle channel for all currency pairs.
+func (p *MexcProvider) subscribeCandles(cps ...types.CurrencyPair) error {
+	pairs := make([]string, len(cps))
+
+	for i, cp := range cps {
+		pairs[i] = currencyPairToMexcPair(cp)
+	}
+
+	return p.subscribePairs(pairs...)
+}
+
+// subscribedPairsToSlice returns the map of subscribed pairs as a slice.
+func (p *MexcProvider) subscribedPairsToSlice() []types.CurrencyPair {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	return types.MapPairsToSlice(p.subscribedPairs)
+}
+
+func (p *MexcProvider) getTickerPrice(key string) (types.TickerPrice, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	ticker, ok := p.tickers[key]
+	if !ok {
+		return types.TickerPrice{}, fmt.Errorf("mexc provider failed to get ticker price for %s", key)
```

Comment:
use error wrapping, and a base error, eg: `ErrKeyNotFound.Wrap...` or `ErrPriceNotFound.Wrap` (not sure if we created them, if not then let's create).